pkg/cmd/util/clientcmd/factory.go

```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 	c.clients[config.GroupVersion.String()] = client
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(f *Factory) ([]unversioned.GroupResource, error) {
```

Comment:
> The factory is a pain to mock. Looks like you want a RESTMapper and a DiscoveryClient. How about just accepting those.

That would also make this easier to use inside the `Run` step, since it won't need a full `Factory`.  Oh, and I really want to refactor that type.